File: services/media-plane/src/g711.cpp
#include "voxdesk/media/g711.hpp"

namespace voxdesk::media {

namespace {

constexpr std::uint8_t kSignBit = 0x80;
constexpr std::uint8_t kQuantMask = 0x0f;
constexpr int kSegShift = 4;
constexpr std::uint8_t kSegMask = 0x70;
constexpr int kMuBias = 0x84;

// Segment upper bounds, in the 16-bit biased-linear domain. Both mu-law and
// A-law use the same table in the canonical reference: the A-law encoder
// offsets the magnitude by -8 before the search, which shifts the segment
// boundaries into the same grid.
const std::int16_t kSegEnd[8] = {0xFF,   0x1FF,  0x3FF,  0x7FF,
                                 0xFFF,  0x1FFF, 0x3FFF, 0x7FFF};

// First table index whose upper bound is >= val, or 8 if val exceeds all.
int Search(int val) {
  for (int i = 0; i < 8; ++i) {
    if (val <= kSegEnd[i]) {
      return i;
    }
  }
  return 8;
}

}  // namespace

std::uint8_t EncodeUlLaw(std::int16_t sample) {
  int magnitude = sample;
  int mask;
  if (magnitude < 0) {
    magnitude = kMuBias - magnitude;
    mask = 0x7F;
  } else {
    magnitude += kMuBias;
    mask = 0xFF;
  }

  const int seg = Search(magnitude);
  if (seg >= 8) {
    // Out of range: return the maximum-magnitude code for this sign.
    return static_cast<std::uint8_t>(0x7F ^ mask);
  }
  const std::uint8_t uval = static_cast<std::uint8_t>(
      (seg << kSegShift) | ((magnitude >> (seg + 3)) & kQuantMask));
  return static_cast<std::uint8_t>(uval ^ mask);
}

std::int16_t DecodeUlLaw(std::uint8_t code) {
  const int u = static_cast<std::uint8_t>(~code);
  int t = ((u & kQuantMask) << 3) + kMuBias;
  t <<= (u & kSegMask) >> kSegShift;
  return static_cast<std::int16_t>((u & kSignBit) ? (kMuBias - t)
                                                  : (t - kMuBias));
}

std::uint8_t EncodeALaw(std::int16_t sample) {
  int magnitude = sample;
  int mask;
  if (magnitude >= 0) {
    mask = 0xD5;  // sign bit = 1 for positive A-law
  } else {
    mask = 0x55;
    magnitude = -magnitude - 8;
  }

  const int seg = Search(magnitude);
  if (seg >= 8) {
    return static_cast<std::uint8_t>(0x7F ^ mask);
  }

  std::uint8_t aval = static_cast<std::uint8_t>(seg << kSegShift);
  if (seg < 2) {
    aval = static_cast<std::uint8_t>(aval | ((magnitude >> 4) & kQuantMask));
  } else {
    aval = static_cast<std::uint8_t>(
        aval | ((magnitude >> (seg + 3)) & kQuantMask));
  }
  return static_cast<std::uint8_t>(aval ^ mask);
}

std::int16_t DecodeALaw(std::uint8_t code) {
  const int a = code ^ 0x55;
  int t = (a & kQuantMask) << 4;
  const int seg = (a & kSegMask) >> kSegShift;
  switch (seg) {
    case 0:
      t += 8;
      break;
    case 1:
      t += 0x108;
      break;
    default:
      t += 0x108;
      t <<= seg - 1;
      break;
  }
  return static_cast<std::int16_t>((a & kSignBit) ? t : -t);
}

std::vector<std::uint8_t> EncodeUlLaw(const std::int16_t* pcm,
                                      std::size_t count) {
  std::vector<std::uint8_t> out(count);
  for (std::size_t i = 0; i < count; ++i) {
    out[i] = EncodeUlLaw(pcm[i]);
  }
  return out;
}

std::vector<std::int16_t> DecodeUlLaw(const std::uint8_t* code,
                                      std::size_t count) {
  std::vector<std::int16_t> out(count);
  for (std::size_t i = 0; i < count; ++i) {
    out[i] = DecodeUlLaw(code[i]);
  }
  return out;
}

std::vector<std::uint8_t> EncodeALaw(const std::int16_t* pcm,
                                     std::size_t count) {
  std::vector<std::uint8_t> out(count);
  for (std::size_t i = 0; i < count; ++i) {
    out[i] = EncodeALaw(pcm[i]);
  }
  return out;
}

std::vector<std::int16_t> DecodeALaw(const std::uint8_t* code,
                                     std::size_t count) {
  std::vector<std::int16_t> out(count);
  for (std::size_t i = 0; i < count; ++i) {
    out[i] = DecodeALaw(code[i]);
  }
  return out;
}

}  // namespace voxdesk::media

File: services/realtime/media-engine-rs/crates/audio/src/g711.rs
//! G.711 µ-law and A-law companding (ITU-T G.711): the payload codec of
//! PSTN-facing legs (RTP payload types 0 and 8).
//!
//! Implemented from the standard reference algorithms (the same arithmetic
//! published by Sun Microsystems as the G.711 reference and reproduced in
//! ITU-T G.191's test material), at a consistent 16-bit operating point:
//!
//! * decode: companded byte → signed 16-bit linear PCM
//! * encode: signed 16-bit linear PCM → companded byte (lossy quantization:
//!   round-trip error stays within the codec's design step size, which the
//!   conformance tests assert rather than pretending bit-exactness exists)
//!
//! Mixing happens in linear PCM (mixer.rs), so these functions are the
//! boundary of every companded leg.

// ---------------------------------------------------------------------------
// µ-law (North America / Japan; RTP payload type 0)
// ---------------------------------------------------------------------------

/// µ-law's bias and linear clip point at the 16-bit operating point, and
/// the segment upper-bound table (0xFF, 0x1FF, … 0x7FFF).
const ULAW_BIAS: i32 = 0x84; // 132
const ULAW_CLIP: i32 = 32635;
const ULAW_SEG_END: [i32; 8] = [0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF, 0x1FFF, 0x3FFF, 0x7FFF];

/// Decode one µ-law byte to 16-bit linear PCM (reference arithmetic:
/// t = ((mantissa << 3) + BIAS) << segment; sample = t - BIAS, signed).
pub fn ulaw_decode(u_val: u8) -> i16 {
    let u = !u_val; // ones' complement per G.711
    let sign = u & 0x80;
    let segment = i32::from((u >> 4) & 0x07);
    let mantissa = i32::from(u & 0x0F);
    let t = (((mantissa << 3) + ULAW_BIAS) << segment) - ULAW_BIAS;
    if sign != 0 {
        -t as i16
    } else {
        t as i16
    }
}

/// Encode one 16-bit linear PCM sample to µ-law (reference arithmetic:
/// bias, clip, segment search, then (seg<<4)|quant with the polarity
/// mask folded in).
pub fn ulaw_encode(sample: i16) -> u8 {
    let s = i32::from(sample);
    let mask: u8;
    let mut pcm = if s < 0 {
        mask = 0x7F;
        ULAW_BIAS - s // note: -s, not s — the bias subtraction is the reference's convention
    } else {
        mask = 0xFF;
        s + ULAW_BIAS
    };
    if pcm > ULAW_CLIP {
        pcm = ULAW_CLIP;
    }
    let segment = ULAW_SEG_END.iter().position(|&end| pcm <= end).unwrap_or(8) as i32;
    if segment >= 8 {
        return 0x7F ^ mask; // out of range: maximum magnitude code
    }
    let quant = ((pcm >> (segment + 3)) & 0x0F) as u8;
    ((segment as u8) << 4 | quant) ^ mask
}

// ---------------------------------------------------------------------------
// A-law (Europe / rest of world; RTP payload type 8)
// ---------------------------------------------------------------------------

/// Segment upper bounds at the 13-bit scale: 0x1F, 0x3F, 0x7F, …
const ALAW_SEG_END: [i32; 8] = [0x1F, 0x3F, 0x7F, 0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF];

/// Decode one A-law byte to 16-bit linear PCM.
pub fn alaw_decode(a_val: u8) -> i16 {
    let a = a_val ^ 0x55; // even-bit inversion per G.711
    let sign = a & 0x80;
    let segment = i32::from((a >> 4) & 0x07);
    let mut t = i32::from(a & 0x0F) << 4; // to 16-bit scale
    match segment {
        0 => t += 8,
        1 => t += 0x108,
        _ => {
            t += 0x108;
            t <<= segment - 1;
        }
    }
    if sign != 0 {
        t as i16
    } else {
        -(t as i16)
    }
}

/// Encode one 16-bit linear PCM sample to A-law.
pub fn alaw_encode(sample: i16) -> u8 {
    let s = i32::from(sample) >> 3; // to the 13-bit scale
    let mask: u8;
    let mut pcm13 = if s >= 0 {
        mask = 0xD5;
        s
    } else {
        mask = 0x55;
        -s - 1
    };
    let segment = match ALAW_SEG_END.iter().position(|&end| pcm13 <= end) {
        Some(seg) => seg as i32,
        None => return 0x7F ^ mask, // out of range: maximum magnitude
    };
    let _ = &mut pcm13;
    let quant = if segment < 2 {
        ((pcm13 >> 1) & 0x0F) as u8
    } else {
        ((pcm13 >> segment) & 0x0F) as u8
    };
    ((segment as u8) << 4 | quant) ^ mask
}

// ---------------------------------------------------------------------------
// Payload-wise helpers (RTP payloads are byte strings, not sample slices)
// ---------------------------------------------------------------------------

pub fn decode_ulaw_payload(payload: &[u8], out: &mut Vec<i16>) {
    out.extend(payload.iter().map(|&b| ulaw_decode(b)));
}

pub fn encode_ulaw_payload(samples: &[i16], out: &mut Vec<u8>) {
    out.extend(samples.iter().map(|&s| ulaw_encode(s)));
}

pub fn decode_alaw_payload(payload: &[u8], out: &mut Vec<i16>) {
    out.extend(payload.iter().map(|&b| alaw_decode(b)));
}

pub fn encode_alaw_payload(samples: &[i16], out: &mut Vec<u8>) {
    out.extend(samples.iter().map(|&s| alaw_encode(s)));
}
